### PR TITLE
Add the ability to get keys and values from KeyFMap

### PR DESCRIPTION
**Pull request by bashor**
Changes in KeyFMap are needed for Kotlin

### CODE REVIEW

**Comment by bashor** [comment]
@donnerpeter @chashnikov please review this.

**Comment by bashor** [comment]
Note I didn't test it on CI

**Comment by donnerpeter** [comment]
Why exactly do you need this in Kotlin?

**Comment by bashor** [comment]
> Why exactly do you need this in Kotlin?
We need lightweight container for storing metadata in some places.
We want to use `UserDataHolder(Base)` but sometimes we need merge some containers, so we should know what keys stored inside container... 

TBH we need only `getKeys`, `getValues` added because it's already exist in most implementations and to be symmetrical :)

**Comment by bashor** [comment]
Another possible way to achieve it is declare some utility functions outside, but I afraid that then we can forget to update these functions while adding new implementations.

Thoughts?

**Comment by bashor** [comment]
I thought that maybe will be useful to have `containsKey` too?
What do you think?

File: platform/util/src/com/intellij/openapi/util/Key.java [hunk 1]
```diff
@@ -106,6 +106,17 @@ public void set(@Nullable Map<Key, Object> holder, T value) {
     return (Key<T>)allKeys.get(index);
   }
 
+  @NotNull
+  public static Key[] getKeysByIndices(int[] indexes) {
```
Comment (chashnikov) [inline platform/util/src/com/intellij/openapi/util/Key.java]: Index of a key is an implementation detail and shouldn't be exposed to open API. It's bad that we already have public 'getKeyByIndex' method, I would prefer to make it package-local. But anyway let's not emphasize this problem by adding implementation-specific utility method to API.

**Comment by chashnikov** [comment]
If you need to merge KeyFMap why not add method 'plus(KeyFMap)' to KeyFMap
instead?

**Comment by bashor** [comment]
>If you need to merge KeyFMap why not add method 'plus(KeyFMap)' to KeyFMap
instead?

Good point, thanks @chashnikov!
What do You think @donnerpeter?

Anyway I think `getKeys` is useful in some cases(e.g. see second commit).

**Comment by bashor** [comment]
I'm sorry, but I jumped to conclusions. We really need `getKeys`.

**Comment by donnerpeter** [comment]
I still don't understand the exact problem you're trying to solve and why just plain old Map isn't enough. KeyFMap is pretty specific and has quite a bad performance when large. If you need immutability, have a look at SmartFMap (and if Kotlin still doesn't have its own immutable collections, that's a shame).

**Comment by bashor** [comment]
I updated the PR

**Comment by bashor** [comment]
> I still don't understand the exact problem you're trying to solve and why just plain old Map isn't enough. 

We want to store metadatas which key type is `Key`.

>KeyFMap is pretty specific and has quite a bad performance when large. 

It's ok for our case -- we don't expected large datas.

**Comment by bashor** [comment]
We (sometimes) prefer to use tested for years things instead of reinvent it :)

**Comment by donnerpeter** [comment]
java.util.HashMap is a very tested thing, for many more years already than KeyFMap. The latter is itself an implementation detail of UserDataHolderBase and might change in future (although not very likely).  It also lacks some functionality that might be important, e.g. equals/hashCode. HashMap is not so type-safe, agreed, but definitely more reliable and proven.

So I'd say I need more arguments that you really-really need KeyFMap and no alternatives will ever suit you :)

**Comment by bashor** [comment]
I updated PR.

**Comment by bashor** [comment]
Please backport changes to 141 branch

**Comment by donnerpeter** [comment]
Pushed to master. Need review for 141. @chashnikov ?

**Comment by chashnikov** [comment]
Looks ok now, though I would generate KEYS and VALUES arrays in KeyFMapTest instead of hardcoding.

**Comment by bashor** [comment]
I created new PR #262 with changes in KeyFMapTest.

Thanks all!

**Pull request merged**